pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package selectors
+
+import (
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+type headTailIterator struct {
+	head, tail []promql.FPoint
+	idx        int // combined iterator position across both
+	lazyLen    int // combined length, lazy initialized in len()
+}
+
+func (i *headTailIterator) len() int {
+	// avoid having to do len(i.head) + len(i.tail) on each call
+	if i.lazyLen == 0 && (len(i.head) > 0 || len(i.tail) > 0) {
+		i.lazyLen = len(i.head) + len(i.tail)
+	}
+	return i.lazyLen
+}
+
+func (i *headTailIterator) hasNext() bool {
+	return i.idx < i.len()
+}
+
+func (i *headTailIterator) next() *promql.FPoint {
+	if i.idx >= i.len() {
+		return nil
+	}
+	p := i.at(i.idx)
+	i.idx++
+	return p
+}
+
+func (i *headTailIterator) prev() *promql.FPoint {
+	if i.idx <= 0 {
+		return nil
+	}
+	i.idx--
+	return i.at(i.idx)
+}
+
+func (i *headTailIterator) inc() {
+	if i.idx < i.len() {
+		i.idx++
+	}
+}
+
+func (i *headTailIterator) peek() *promql.FPoint {
+	if i.idx >= i.len() {
+		return nil
+	}
+	return i.at(i.idx)
+}
+
+func (i *headTailIterator) at(pos int) *promql.FPoint {
+	if pos < len(i.head) {
+		return &i.head[pos]
+	}
+	return &i.tail[pos-len(i.head)]
+}
+
+// scanTo will return the point which is closest to being <= time, or the first point after this time.
+// The iterator will be positioned to the first point > time.
+func (i *headTailIterator) scanTo(time int64) *promql.FPoint {
+	var first *promql.FPoint
+
+	for i.hasNext() {
+		next := i.peek()
+
+		if next.T < time {
+			first = next
+			i.inc()
+			continue
+		}
+
+		if next.T == time {
+			i.inc()
+			return next
+		}
+
+		if first != nil {
+			return first
+		}
+
+		return next
+	}
+
+	return first
+}
+
+// accumulateTo will accumulate all points <= time into the given buff.
+// The first point which is >= time is returned, and if not found the last point < time is returned.
```

Comment:
Returning the next point from a method named `accumulateTo` seems out of place - perhaps we could not return it, and instead rely on the caller calling `at` to get it?